pilot/pkg/networking/plugin/authz/rbac_v2.go

```diff
@@ -0,0 +1,129 @@
+// * Deprecate ServiceRoleBinding. Only support two CRDs: ServiceRole and AuthorizationPolicy.
+// * Allow multiple bindings and roles in one CRD, i.e. Authorization.
+// * Support workload selector.
+package authz
+
+import (
+	"fmt"
+
+	http_config "github.com/envoyproxy/go-control-plane/envoy/config/filter/http/rbac/v2"
+	policyproto "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v2alpha"
+
+	rbacproto "istio.io/api/rbac/v1alpha1"
+	"istio.io/istio/pilot/pkg/model"
+)
+
+// convertRbacRulesToFilterConfigV2 is the successor of convertRbacRulesToFilterConfig, which supports
+// converting AuthorizationPolicy.
+func convertRbacRulesToFilterConfigV2(service *serviceMetadata, option rbacOption) *http_config.RBAC {
+	rbac := &policyproto.RBAC{
+		Action:   policyproto.RBAC_ALLOW,
+		Policies: map[string]*policyproto.Policy{},
+	}
+
+	if option.globalPermissiveMode {
+		// Permissive Mode will be ignored.
+		rbacLog.Debugf("Permissive mode is not supported for service %s.", service.name)
+		// TODO(pitlv2109): Handle permissive mode in the future.
+		return &http_config.RBAC{Rules: rbac}
+	}
+
+	namespace := service.attributes[attrDestNamespace]
+	if option.authzPolicies == nil || option.authzPolicies.NamespaceToAuthorizationConfigV2 == nil {
+		return &http_config.RBAC{Rules: rbac}
+	}
+	if _, present := option.authzPolicies.NamespaceToAuthorizationConfigV2[namespace]; !present {
+		return &http_config.RBAC{Rules: rbac}
+	}
+	// Get all AuthorizationPolicy Istio config from this namespace.
+	allAuthzPolicies := option.authzPolicies.NamespaceToAuthorizationConfigV2[namespace].AuthzPolicies
+	for _, authzPolicy := range allAuthzPolicies {
+		// If a WorkloadSelector is used in the AuthorizationPolicy config and does not match this service, skip this
+		// AuthorizationPolicy.
+		serviceLabelsCollection := model.LabelsCollection{service.labels}
+		if authzPolicy.Policy.WorkloadSelector != nil &&
+			!(serviceLabelsCollection.IsSupersetOf(authzPolicy.Policy.WorkloadSelector.Labels)) {
+			continue
+		}
+		for i, binding := range authzPolicy.Policy.Allow {
+			// Check if we have converted this ServiceRole before.
+			permissions := make([]*policyproto.Permission, 0)
+			// Convert role.
+			role := option.authzPolicies.RoleForNameAndNamespace(namespace, binding.RoleRef.Name)
+			for _, rule := range role.Rules {
+				// Check to make sure the services field does not exist in the ServiceRole.
+				if len(rule.Services) > 0 {
+					rbacLog.Errorf("should not have services field in ServiceRole (found in %s)", binding.RoleRef.Name)
+					continue
+				}
+				// Check to make sure that the current service (caller) is the one this rule is applying to.
+				if !service.areConstraintsMatched(rule) {
+					rbacLog.Errorf("rule has constraints but doesn't match with the service (found in %s)", binding.RoleRef.Name)
+					continue
+				}
+				if option.forTCPFilter {
+					// TODO(yangminzhu): Move the validate logic to push context and add metrics.
```

Comment:
Is this TODO still valid? @yangminzhu

Reply:
Part of it, we can add metrics, but I think the validate logic cannot be moved to push context.

Reply:
Thanks @yangminzhu. Let's update the comments here.

Reply:
I can update the comment for @yangminzhu in this PR.